storage/postgres/client.go

```diff
@@ -96,89 +95,99 @@ func (c *Client) SendBatch(ctx context.Context, batch *storage.QueryBatch) error
-					}
+		batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+	} else {
+		// use implicit tx provided by SendBatch; see https://github.com/jackc/pgx/issues/879
+		batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+	}
+	defer common.CloseOrLog(batchResults, c.logger)
+
+	// Exec indiviual queries in the batch.
+	for i := 0; i < pgxBatch.Len(); i++ {
+		if _, err := batchResults.Exec(); err != nil {
+			rollbackErr := ""
+			if useExplicitTx {
+				err2 := tx.Rollback(ctx)
+				if err2 != nil {
+					rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err2.Error())
 				}
-				return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
 			}
+			return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
 		}
+	}
 
-		// Commit the tx.
-		if useExplicitTx {
-			err := tx.Commit(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to commit tx: %w", err)
-			}
-		}
-	} else {
-		// Begin a transaction.
-		tx, err := c.pool.BeginTx(ctx, opts)
+	// Commit the tx.
+	if useExplicitTx {
+		err := tx.Commit(ctx)
 		if err != nil {
-			return fmt.Errorf("failed to begin tx: %w", err)
+			return fmt.Errorf("failed to commit tx: %w", err)
 		}
+	}
+	return nil
+}
 
-		// Exec indiviual queries in the batch.
-		for i, q := range batch.Queries() {
-			if _, err2 := tx.Exec(ctx, q.Cmd, q.Args...); err2 != nil {
-				rollbackErr := ""
-				err3 := tx.Rollback(ctx)
-				if err3 != nil {
-					rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err3.Error())
-				}
-				return fmt.Errorf("query %d %v: %w%s", i, q, err2, rollbackErr)
+// Submits a new batch of queries, sending one query at a time. Compared with `sendBatchWithOptionsSlow`, this
+// gives slower performance but better error reporting.
+func (c *Client) sendBatchWithOptionsSlow(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	// Begin a transaction.
+	tx, err := c.pool.BeginTx(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("failed to begin tx: %w", err)
+	}
+
+	// Exec indiviual queries in the batch.
+	for i, q := range batch.Queries() {
+		if _, err2 := tx.Exec(ctx, q.Cmd, q.Args...); err2 != nil {
+			rollbackErr := ""
+			err3 := tx.Rollback(ctx)
+			if err3 != nil {
+				rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err3.Error())
 			}
+			return fmt.Errorf("query %d %v: %w%s", i, q, err2, rollbackErr)
 		}
+	}
 
-		// Commit the transaction.
-		err = tx.Commit(ctx)
-		if err != nil {
-			c.logger.Error("failed to submit tx",
-				"error", err,
-				"batch", batch.Queries(),
-			)
-			return err
-		}
+	// Commit the transaction.
+	err = tx.Commit(ctx)
+	if err != nil {
+		c.logger.Error("failed to submit tx",
+			"error", err,
+			"batch", batch.Queries(),
+		)
+		return err
 	}
 	return nil
 }
 
+func (c *Client) SendBatchWithOptions(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	if err := c.sendBatchWithOptionsFast(ctx, batch, opts); err == nil {
```

Comment:
shouldn't throw away `err` here imo, we might not get the same error or any error from -Slow

Reply:
It wasn't thrown away, it was logged inside the Fast() method. But that was not a logical place for it :+1:.
Moved the logging here now.